### ISSUE TITLE
Parse error  - different base classes

### ISSUE THREAD

**Issue by amai2012** [SonarOpenCommunity/sonar-cxx#2336]
**Describe the bug**
Unexpected parse error

**To Reproduce**
Either run analysis or use sslr tool

**Expected behavior**
No error :-)

**Screenshots**

**Desktop (please complete the following information):**
 - OS: any (currently Windows...)
 - SonarQube version: 8.9.x LTS
 - cxx plugin version: 2.0.7
 - sonar-scanner version: [e.g. 4.6.0.2311]
 
**Additional context**

```C++
#if !defined AAAAAAAAAAAAAAAAAAAAAAAAAA
#define AAAAAAAAAAAAAAAAAAAAAAAAAA

#ifdef SWIG
class Foo: public Bar
#else
class Foo : public MyTemplate<Bar>, public YXX
#endif
{
    FooPtr createInstance() const override;

};

#endif 
```

**Comment by guwirth** [comment]
Hi @amai2012,

we will have a look to it but can take somehow longer. At the moment we prepare v2.1.

Regards,

**Comment by guwirth** [comment]
Hi @amai2012,

you can help providing the stack trace from the LOG file from the exception.

Regards,

**Comment by amai2012** [comment]
@guwirth  Here it is:
```
java.lang.NullPointerException
	at org.sonar.sslr.internal.toolkit.ToolkitViewImpl.getCodeElementStartOffset(ToolkitViewImpl.java:349)
	at org.sonar.sslr.internal.toolkit.ToolkitViewImpl.getAstNodeFollowingCurrentSourceCodeTextCursorPosition(ToolkitViewImpl.java:471)
	at org.sonar.sslr.internal.toolkit.ToolkitPresenter.onSourceCodeTextCursorMoved(ToolkitPresenter.java:169)
	at org.sonar.sslr.internal.toolkit.ToolkitViewImpl$3.caretUpdate(ToolkitViewImpl.java:184)
	at java.desktop/javax.swing.text.JTextComponent.fireCaretUpdate(JTextComponent.java:412)
	at java.desktop/javax.swing.text.JTextComponent$MutableCaretEvent.fire(JTextComponent.java:4489)
	at java.desktop/javax.swing.text.JTextComponent$MutableCaretEvent.stateChanged(JTextComponent.java:4511)
	at java.desktop/javax.swing.text.DefaultCaret.fireStateChanged(DefaultCaret.java:812)
	at java.desktop/javax.swing.text.DefaultCaret.changeCaretPosition(DefaultCaret.java:1283)
	at java.desktop/javax.swing.text.DefaultCaret.handleSetDot(DefaultCaret.java:1182)
	at java.desktop/javax.swing.text.DefaultCaret.setDot(DefaultCaret.java:1163)
	at java.desktop/javax.swing.text.DefaultCaret$Handler.removeUpdate(DefaultCaret.java:1831)
	at java.desktop/javax.swing.text.AbstractDocument.fireRemoveUpdate(AbstractDocument.java:261)
	at java.desktop/javax.swing.text.html.HTMLDocument.access$300(HTMLDocument.java:275)
	at java.desktop/javax.swing.text.html.HTMLDocument$HTMLReader.adjustEndElement(HTMLDocument.java:2692)
	at java.desktop/javax.swing.text.html.HTMLDocument$HTMLReader.flush(HTMLDocument.java:2731)
	at java.desktop/javax.swing.text.html.HTMLEditorKit.read(HTMLEditorKit.java:264)
	at java.desktop/javax.swing.plaf.basic.BasicTextUI$TextTransferHandler.handleReaderImport(BasicTextUI.java:2398)
	at java.desktop/javax.swing.plaf.basic.BasicTextUI$TextTransferHandler.importData(BasicTextUI.java:2602)
	at java.desktop/javax.swing.TransferHandler.importData(TransferHandler.java:827)
	at java.desktop/javax.swing.plaf.basic.BasicTextUI$TextTransferHandler.importData(BasicTextUI.java:2540)
	at java.desktop/javax.swing.TransferHandler$TransferAction.actionPerformedImpl(TransferHandler.java:1763)
	at java.desktop/javax.swing.TransferHandler$TransferAction$1.run(TransferHandler.java:1712)
	at java.desktop/javax.swing.TransferHandler$TransferAction$1.run(TransferHandler.java:1710)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:389)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:95)
	at java.desktop/javax.swing.TransferHandler$TransferAction$2.run(TransferHandler.java:1727)
	at java.desktop/javax.swing.TransferHandler$TransferAction$2.run(TransferHandler.java:1725)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:389)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
	at java.desktop/javax.swing.TransferHandler$TransferAction.actionPerformed(TransferHandler.java:1724)
	at java.desktop/javax.swing.text.JTextComponent.invokeAction(JTextComponent.java:1523)
	at java.desktop/javax.swing.text.JTextComponent.paste(JTextComponent.java:1501)
	at java.desktop/javax.swing.text.DefaultEditorKit$PasteAction.actionPerformed(DefaultEditorKit.java:1373)
	at java.desktop/javax.swing.SwingUtilities.notifyAction(SwingUtilities.java:1810)
	at java.desktop/javax.swing.JComponent.processKeyBinding(JComponent.java:2900)
	at java.desktop/javax.swing.JComponent.processKeyBindings(JComponent.java:2948)
	at java.desktop/javax.swing.JComponent.processKeyEvent(JComponent.java:2862)
	at java.desktop/java.awt.Component.processEvent(Component.java:6409)
	at java.desktop/java.awt.Container.processEvent(Container.java:2263)
	at java.desktop/java.awt.Component.dispatchEventImpl(Component.java:5008)
	at java.desktop/java.awt.Container.dispatchEventImpl(Container.java:2321)
	at java.desktop/java.awt.Component.dispatchEvent(Component.java:4840)
	at java.desktop/java.awt.KeyboardFocusManager.redispatchEvent(KeyboardFocusManager.java:1950)
	at java.desktop/java.awt.DefaultKeyboardFocusManager.dispatchKeyEvent(DefaultKeyboardFocusManager.java:870)
	at java.desktop/java.awt.DefaultKeyboardFocusManager.preDispatchKeyEvent(DefaultKeyboardFocusManager.java:1139)
	at java.desktop/java.awt.DefaultKeyboardFocusManager.typeAheadAssertions(DefaultKeyboardFocusManager.java:1009)
	at java.desktop/java.awt.DefaultKeyboardFocusManager.dispatchEvent(DefaultKeyboardFocusManager.java:835)
	at java.desktop/java.awt.Component.dispatchEventImpl(Component.java:4889)
	at java.desktop/java.awt.Container.dispatchEventImpl(Container.java:2321)
	at java.desktop/java.awt.Window.dispatchEventImpl(Window.java:2762)
	at java.desktop/java.awt.Component.dispatchEvent(Component.java:4840)
	at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:772)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:721)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:715)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:389)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:95)
	at java.desktop/java.awt.EventQueue$5.run(EventQueue.java:745)
	at java.desktop/java.awt.EventQueue$5.run(EventQueue.java:743)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:389)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
	at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:742)
	at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:203)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:124)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:113)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:109)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:90)
``

**Comment by amai2012** [comment]
BTW, I think this could fix the last syntax errors in my project for now. There are still ~ 100, but all in just 2 files which look quite similar.

In addition I had to add additional macros (things like __declspec(dllexport) wrapped inside other macros)

**Comment by guwirth** [comment]
> In addition I had to add additional macros (things like __declspec(dllexport) wrapped inside other macros)

@amai2012 that's by intention:
https://github.com/SonarOpenCommunity/sonar-cxx/wiki/Dealing-with-compiler-specific-code-pieces

**Comment by guwirth** [comment]
@amai2012 you can try with https://github.com/SonarOpenCommunity/sonar-cxx/releases/tag/latest-snapshot
